### PR TITLE
Implement metrics

### PR DESCRIPTION
**Pull request by roby2014**
# Description

**Check the sample for usage**

TLDR: The goal is to have a simple but extendible core class to register, search and debug metrics, with a compiler option to enable it

## Checklist

- [X] Self-review changes.
- [X] Evaluate impact on the documentation.
- [x] Ensure test coverage.
- [X] Write new samples.
- [x] Add entry to the changelog's unreleased section.

### CODE REVIEW

**Comment by RiscadoA** [comment]
Since this PR is still blocked on the other one, I'm marking it as a draft. Does it really depend on the other PR though? They seem mostly unrelated.

**Comment by roby2014** [comment]
It does not depend anymore yeah, you're right.

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 
 - UI canvas, elements, and color rects. (#1116, **@DiogoMendonc-a**)
+- Implemented file-based logging (#1154, **@roby2014**).
```
Comment (RiscadoA) [inline CHANGELOG.md:13]: This shouldn't be here, right?

File: core/include/cubos/core/metrics.hpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        std::string mName; ///< The name of the metric.
+        double mValue;     ///< The value of the metric (as a double).
+        std::string mFile; ///< The source file where the metric was created.
+        int mLine;         ///< The line number where the metric was created.
+    };
+
+    /// @brief Singleton class that manages a collection of metrics.
+    /// @ingroup core
+    class Metrics
```
Comment (RiscadoA) [inline core/include/cubos/core/metrics.hpp:25]: Missing CUBOS_CORE_API

File: core/include/cubos/core/metrics.hpp [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        /// @param seenCount Seen metrics count.
+        /// @return Whether a new metric was found.
+        static bool readName(std::string& name, size_t& seenCount);
+    };
+
+    /// @brief Utility to measure the execution time of a scope.
+    ///
+    /// This class measures the time between its construction and destruction,
+    /// providing a way to profile code execution time.
+    class TimeTracker
```
Comment (RiscadoA) [inline core/include/cubos/core/metrics.hpp:57]: ScopeProfiler perhaps would be a better name? TimeTracker seems unrelated to the metrics system

File: core/include/cubos/core/metrics.hpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+    private:
+        std::string mName;                                 ///< The name of the TimeTracker.
+        std::chrono::_V2::system_clock::time_point mStart; ///< Start time when constructed.
+    };
+} // namespace cubos::core
+
+/// @brief Macro to add a metric.
+/// @param name The name of the metric.
+/// @param val The value to set for the metric.
```
Comment (RiscadoA) [inline core/include/cubos/core/metrics.hpp:74]: Missing @ ingroup core

File: core/include/cubos/core/metrics.hpp [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+#define CUBOS_METRIC(name, val) ::cubos::core::Metrics::metric({name, val, __FILE__, __LINE__})
+
+/// @brief Macro to measure the execution time of a function/scope, registering it as a metric.
+/// @note The macro uses `__FUNCTION__` to register the metric name.
```
Comment (RiscadoA) [inline core/include/cubos/core/metrics.hpp:78]: Missing in group

File: core/samples/metrics/main.cpp [hunk 6]
```diff
@@ -0,0 +1,75 @@
+/// [Include metrics]
+#include <cubos/core/metrics.hpp>
+/// [Include metrics]
+#include <chrono>
+#include <thread>
```
Comment (RiscadoA) [inline core/samples/metrics/main.cpp:5]: I dont think you need those includes

File: core/src/metrics.cpp [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+size_t Metrics::size()
+{
+    return state().mMetrics.size();
+}
+
+bool Metrics::readValue(const std::string& name, double& value, size_t& offset)
+{
+    std::lock_guard<std::mutex> lock(state().mMutex);
+
+    std::vector<Metric> filteredMetricsByName;
```
Comment (RiscadoA) [inline core/src/metrics.cpp:47]: Why not store a map from metric names to vectors in the singleton?
This way you wouldnt need this expensive operation here

File: core/include/cubos/core/metrics.hpp [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+#define CUBOS_PROFILE() ::cubos::core::TimeTracker _profilingTimeTracker(__FUNCTION__)
```
Comment (RiscadoA) [inline core/include/cubos/core/metrics.hpp:79]: Probably a good idea to also add the file and line of code.

**Review by RiscadoA — CHANGES_REQUESTED** [review]
Thanks for working on this :)

File: core/samples/metrics/main.cpp (hunk 6 above)
Comment (roby2014) [reply core/samples/metrics/main.cpp:5]: I do: `std::this_thread::sleep_for(std::chrono::seconds(2));`

Comment (roby2014) [reply core/samples/metrics/main.cpp:5]: at least, the `thread` one, `chrono` can be removed

Comment (RiscadoA) [reply core/samples/metrics/main.cpp]: Ah right, mb

File: core/samples/metrics/main.cpp [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    CUBOS_ASSERT(offset == 1);
+
+    // etc..
+    /// [Getting values]
+
+    /// [Profiling metrics 2]
+    myFunction();
+
+    offset = 0;
+    CUBOS_ASSERT(Metrics::readValue("myFunction", value, offset));
```
Comment (RiscadoA) [inline core/samples/metrics/main.cpp:72]: Won't this fail now, due to __FILE__ + __LINE__? Maybe you can get the name through the `readName` function.

File: core/src/metrics.cpp [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+    std::lock_guard<std::mutex> lock(state().mMutex);
+
+    if (seenCount >= state().mMetrics.size())
+    {
+        return false;
+    }
+
+    // std::map does not support get by index
+    // so we manually loop it like this...
+    for (std::size_t idx = 0; const auto& n : state().mMetrics)
```
Comment (RiscadoA) [inline core/src/metrics.cpp:126]: This is a problem, as the map doesn't have a fixed order. It sorts its keys by lexicographical roder.
Thus, you could end up missing new names and getting old ones repeated when calling this function.
Instead of doing this, keep a separate `vector` for the known metric names.
Each time you add a metric in `::metric` with a name which isn't on the map, you push it to the vector.
Then, here you just access that vector.

**Review by RiscadoA — CHANGES_REQUESTED** [review]

File: core/src/metrics.cpp [hunk 11]
```diff
@@ -0,0 +1,163 @@
+#include <deque>
+#include <map>
+#include <utility>
+
+#include <cubos/core/log.hpp>
+#include <cubos/core/metrics.hpp>
+#include <cubos/core/reflection/external/primitives.hpp>
```
Comment (RiscadoA) [inline core/src/metrics.cpp:7]:
```suggestion
#include <cubos/core/reflection/external/primitives.hpp>
#include <cubos/core/reflection/external/string.hpp>
```

File: core/include/cubos/core/metrics.hpp [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+namespace cubos::core
+{
+    /// @brief Represents a metric.
+    struct Metric
+    {
+        CUBOS_REFLECT;
+        std::string mName; ///< The name of the metric.
+        double mValue;     ///< The value of the metric (as a double).
+        std::string mFile; ///< The source file where the metric was created.
+        std::size_t mLine; ///< The line number where the metric was created.
```
Comment (RiscadoA) [inline core/include/cubos/core/metrics.hpp:24]: I don't think this matters for generic metrics. Just the name should be enough.
On the profiler side, you can add this information to the name of the metric - this is necessary as to not have conflicts between functions with the same name.

File: core/include/cubos/core/metrics.hpp [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+#include <cubos/core/api.hpp>
+#include <cubos/core/reflection/reflect.hpp>
+
+namespace cubos::core
+{
+    /// @brief Represents a metric.
+    struct Metric
+    {
+        CUBOS_REFLECT;
+        std::string mName; ///< The name of the metric.
```
Comment (RiscadoA) [inline core/include/cubos/core/metrics.hpp:21]: At this point you don't need to store this anymore, as the values are stored in a map with names as keys anyways.
Thus, you can just keep a `std::deque` of doubles, which should be faster. Which means that, maybe there's no need at all to have this type.

**Review by RiscadoA — CHANGES_REQUESTED** [review]
Sorry for being so nitpicky with these reviews! Since this is something we'll start using everywhere I really want to get the details right. Ty for the all the work :pray:

File: core/include/cubos/core/metrics.hpp (hunk 12 above)
Comment (roby2014) [reply core/include/cubos/core/metrics.hpp:24]: Hmmm, like `myFunction:54:main.cpp` ? The problem will be reading the metric...

Comment (RiscadoA) [reply core/include/cubos/core/metrics.hpp:24]: You would never want to do that though, you would find its name through `readName`.
Btw, we need a way to distinguish profiling metrics from other metrics, maybe we should prefix their name?

Comment (roby2014) [reply core/include/cubos/core/metrics.hpp:24]: `prof__myFunction` and `metric__a` looks good?

Comment (roby2014) [reply core/include/cubos/core/metrics.hpp:24]: IDK, thinking about it later on the UI, not sure how it should look.

Comment (RiscadoA) [reply core/include/cubos/core/metrics.hpp:24]: On the UI we would probably filter out those and give them special treatment. Maybe we should some special character which doesn't appear in function names? For example, `profiling:<file>:<function name:/line number`

Comment (roby2014) [reply core/include/cubos/core/metrics.hpp:24]: so for normal metrics, no prefix, correct?

**Review by RiscadoA — APPROVED** [review]
LGTM! :ship:

File: core/include/cubos/core/metrics.hpp [hunk 14]
```diff
@@ -0,0 +1,107 @@
+/// @file
+/// @brief Metrics and profiling utilities and macros.
+/// @ingroup core
+
+#pragma once
+
+#include <algorithm>
```
Comment (RiscadoA) [inline core/include/cubos/core/metrics.hpp:7]:
```suggestion
```

Just noticed this, do not include this in headers! It eats up compile times because it includes pretty much the whole STL

File: core/include/cubos/core/metrics.hpp [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+#include <chrono>
+#include <mutex>
```
Comment (RiscadoA) [inline core/include/cubos/core/metrics.hpp:9]:
```suggestion
```
You only use this in the source file, move it there

**Comment by roby2014** [comment]
Eh no idea why static analysis is failing

**Comment by RiscadoA** [comment]
> Eh no idea why static analysis is failing

It's an error in the action. It's trying to dismiss a review which no longer exists and crashes.
TL;DR: not a problem in your code. @joaomanita @diogomsmiranda pls review when you have the chance

File: core/CMakeLists.txt [hunk 16]
```diff
@@ -17,6 +17,9 @@ option(BUILD_CORE_SAMPLES "Build cubos core samples" OFF)
 option(BUILD_CORE_TESTS "Build cubos core tests?" OFF)
 option(BUILD_CORE_SHARED "Build cubos core as shared library?" ON)
 
+option(ENABLE_PROFILING "Should profiling macros be compiled?" ON)
+set(CUBOS_METRIC_MAX_ENTRIES "1024" CACHE STRING "Maximum number of entries each metric can store")
```
Comment (diogomsmiranda) [inline core/CMakeLists.txt:21]: why was 1024 chosen as the max number of entries? Is this arbitrary?

**Review by diogomsmiranda — APPROVED** [review]

**Review by joaomanita — APPROVED** [review]

**Comment by RiscadoA** [comment]
> I also didn't understand the need to make such a sample basically copying from the test. Shouldn't we make something more in tune with what would be a real example?

Agree, we should avoid this, could you improve it @roby2014?

File: core/CMakeLists.txt (hunk 16 above)
Comment (roby2014) [reply core/CMakeLists.txt:21]: Just a dummy value. It can be changed easily through the method `setMaxEntries`

**Comment by diogomsmiranda** [comment]
I would rather have the metrics to be explicitly generated with different names and contexts. But if @RiscadoA is ok with a sample like this It's not a deal breaker for me either.

**Comment by roby2014** [comment]
> I would rather have the metrics to be explicitly generated with different names and contexts. But if @RiscadoA is ok with a sample like this It's not a deal breaker for me either.

Can you give me an example? What should I sample?

**Comment by diogomsmiranda** [comment]
> I also didn't understand the need to make such a sample basically copying from the test. Shouldn't we make something more in tune with what would be a real example?

@roby2014 thank you for addressing this! LGTM

**Comment by RiscadoA** [comment]
You can ignore the failing lint check btw
